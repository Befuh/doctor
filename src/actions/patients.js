export const updatePatients = patients => ({
  type: 'UPDATE_PATIENTS',
  patients
});
