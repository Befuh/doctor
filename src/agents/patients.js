import axios from 'axios';

export const get = async ({ firstName = '', lastName = '', identifier = '', sex = '' }) => {
  if (!firstName && !lastName && !identifier) {
    return;
  }

  let queryParameters = [];
  if (firstName) {
    queryParameters = [...queryParameters, `first_name=${firstName}`];
  }
  if (lastName) {
    queryParameters = [...queryParameters, `last_name=${lastName}`];
  }
  if (identifier) {
    queryParameters = [...queryParameters, `identifier=${identifier}`];
  }
  if (sex) {
    queryParameters = [...queryParameters, `sex=${sex}`];
  }
  const query = queryParameters.join('&');
  const response = await axios.get(`${process.env.BACKEND_URL}/patients?${query}`);

  return response.data.data;
};

export const update = async patient => {
  let url = `${process.env.BACKEND_URL}/patients/${patient.identifier}`;
  const response = await axios.put(url, patient);

  return response.data.data;
};
