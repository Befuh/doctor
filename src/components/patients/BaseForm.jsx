import PropTypes from 'prop-types';
import React, { Component } from 'react';
import { Button, Form, FormGroup, Input, Label } from 'reactstrap';

class BaseForm extends Component {
  static propTypes = {
    patient: PropTypes.object,
    onSubmit: PropTypes.func.isRequired
  };

  state = {
    patient: this.props.patient
  };

  render () {
    const {
      first_name: firstName,
      last_name: lastName,
      date_of_birth: dateOfBirth,
      sex
    } = this.state.patient.user.data;
    return (
      <Form onSubmit={this.submit}>
        <FormGroup>
          <Label for='first_name'>First Name</Label>
          <Input
            type='text'
            id="first_name"
            placeholder='First Name'
            value={firstName}
            onChange={this.changeValue('first_name')}
          />
        </FormGroup>
        <FormGroup>
          <Label for='last_name'>Last Name</Label>
          <Input
            type='text'
            id="last_name"
            placeholder='Last Name'
            value={lastName}
            onChange={this.changeValue('last_name')}
          />
        </FormGroup>
        <FormGroup>
          <Label for='date_of_birth'>Date of Birth</Label>
          <Input
            type='date'
            id="date_of_birth"
            placeholder='Date of Birth'
            value={dateOfBirth}
            onChange={this.changeValue('date_of_birth')}
          />
        </FormGroup>
        <FormGroup>
          <Label for='sex'>Sex</Label>
          <Input type='select' id="sex" placeholder='Sex' value={sex} onChange={this.changeValue('sex')}>
            <option>female</option>
            <option>male</option>
          </Input>
        </FormGroup>
        <div className='text-right'>
          <Button color='primary'>Save Patient</Button>
        </div>
      </Form>
    );
  }

  submit = (event) => {
    event.preventDefault();
    this.props.onSubmit(this.state.patient);
  };

  changeValue = (key) => (event) => {
    event.preventDefault();
    let patient = this.state.patient;
    patient.user.data[key] = event.target.value;
    this.setState(() => ({ patient }));
  };
}

export default BaseForm;
