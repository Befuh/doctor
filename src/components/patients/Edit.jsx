import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import React, { Component, Fragment } from 'react';
import { Button, Form, FormGroup, Input, Label } from 'reactstrap';

export class Edit extends Component {
  propTypes = {
    patient: PropTypes.object.isRequired
  };

  state = {
    patient: this.props.patient
  };

  render () {
    const {
      first_name: firstName,
      last_name: lastName,
      date_of_birth: dateOfBirth,
      sex
    } = this.state.patient.user.data;
    return (
      <Fragment>
        <div className='title'>
          <h1>Edit Patient</h1>
        </div>
        <Form>
          <FormGroup>
            <Label for='first_name'>First Name</Label>
            <Input type='text' id="first_name" placeholder='First Name' value={firstName}/>
          </FormGroup>
          <FormGroup>
            <Label for='last_name'>Last Name</Label>
            <Input type='text' id="last_name" placeholder='Last Name' value={lastName}/>
          </FormGroup>
          <FormGroup>
            <Label for='date_of_birth'>Date of Birth</Label>
            <Input type='date' id="date_of_birth" placeholder='Date of Birth' value={dateOfBirth}/>
          </FormGroup>
          <FormGroup>
            <Label for='sex'>Sex</Label>
            <Input type='select' id="sex" placeholder='Sex' value={sex}>
              <option>female</option>
              <option>male</option>
            </Input>
          </FormGroup>
          <div className='text-right'>
            <Button color='primary'>Save Patient</Button>
          </div>
        </Form>
      </Fragment>
    )
  }
}

const mapStateToProps = () => ({});
const mapDispatchToProps = () => ({});

export default connect(mapStateToProps, mapDispatchToProps)(Edit);
