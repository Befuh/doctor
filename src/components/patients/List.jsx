import React, { PureComponent } from 'react';
import { ButtonGroup, ButtonToolbar, Table } from 'reactstrap';

import './List.scss';
import { FaEye, FaPen } from "react-icons/fa";

class List extends PureComponent {
  render () {
    return (
      <Table striped borderless hover className='patients-list'>
        <thead>
          <tr>
            <th>Identifier</th>
            <th>First Name</th>
            <th>Last Name</th>
            <th>Sex</th>
            <th>Actions</th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <th scope='row'>6tdaf1</th>
            <td>Cristiano</td>
            <td>Ronaldo</td>
            <td>Male</td>
            <td className='actions'>
              <a href='#' className='btn btn-outline-primary btn-sm'><FaEye/> View</a>
              <a href='#' className='btn btn-outline-primary btn-sm'><FaPen/> Edit</a>
            </td>
          </tr>
          <tr>
            <th scope='row'>6tdaf1</th>
            <td>Cristiano</td>
            <td>Ronaldo</td>
            <td>Male</td>
            <td className='actions'>
              <a href='#' className='btn btn-outline-primary btn-sm'><FaEye/> View</a>
              <a href='#' className='btn btn-outline-primary btn-sm'><FaPen/> Edit</a>
            </td>
          </tr>
          <tr>
            <th scope='row'>6tdaf1</th>
            <td>Cristiano</td>
            <td>Ronaldo</td>
            <td>Male</td>
            <td className='actions'>
              <a href='#' className='btn btn-outline-primary btn-sm'><FaEye/> View</a>
              <a href='#' className='btn btn-outline-primary btn-sm'><FaPen/> Edit</a>
            </td>
          </tr>
          <tr>
            <th scope='row'>6tdaf1</th>
            <td>Cristiano</td>
            <td>Ronaldo</td>
            <td>Male</td>
            <td className='actions'>
              <a href='#' className='btn btn-outline-primary btn-sm'><FaEye/> View</a>
              <a href='#' className='btn btn-outline-primary btn-sm'><FaPen/> Edit</a>
            </td>
          </tr>
        </tbody>
      </Table>
    )
  }
}

export default List;
