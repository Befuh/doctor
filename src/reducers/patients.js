const defaultState = {
  list: [],
  searching: false
};

export default (state = defaultState, action) => {
  switch (action.type) {
    case 'UPDATE_PATIENTS':
      const { patients } = action;
      return { ...state, patients };
    case 'START_PATIENT_SEARCH':
      return { ...state, searching: true };
    case 'FINISH_PATIENT_SEARCH':
      return { ...state, searching: false };
    default:
      return state;
  }
};
