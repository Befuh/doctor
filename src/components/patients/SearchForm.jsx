import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import React  from 'react';
import { startUpdatePatients } from '../../actions/patients';

export class SearchForm extends React.Component {
  state = {
    firstName: '',
    lastName: '',
    identifier: '',
    sex: ''
  };

  submit = event => {
    event.preventDefault();
    this.props.startUpdatePatients({
      firstName: this.state.firstName,
      lastName: this.state.lastName,
      identifier: this.state.identifier,
      sex: this.state.sex
    });
  };

  changeInputField = stateName => event => {
    event.preventDefault();
    this.setState({ [stateName]: event.target.value });
  };

  render () {
    return (
      <form onSubmit={this.submit}>
        <input
          onChange={this.changeInputField('firstName')}
        />
        <input
          onChange={this.changeInputField('lastName')}
        />
        <input
          onChange={this.changeInputField('identifier')}
        />
        <select
          onChange={this.changeInputField('sex')}
        >
          <option value='female'>Female</option>
          <option value='male'>male</option>
        </select>
        <button type='submit'>Search</button>
      </form>
    )
  }
}

SearchForm.propTypes = {
  startUpdatePatients: PropTypes.func.isRequired
};

const mapDispatchToProps = dispatch => ({
  startUpdatePatients: query => dispatch(startUpdatePatients(query))
});

export default connect(undefined, mapDispatchToProps)(SearchForm);
