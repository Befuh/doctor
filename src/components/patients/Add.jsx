import BaseForm from './BaseForm';
import PropTypes from 'prop-types';
import React, { Fragment, PureComponent } from 'react';

export class Add extends PureComponent {
  static propTypes = {
    createPatient: PropTypes.func.isRequired
  };

  render () {
    return (
      <Fragment>
        <div className='title'>
          <h1>Add Patient</h1>
        </div>
        <BaseForm onSubmit={this.submit}/>
      </Fragment>
    )
  }

  submit = patient => {
    this.props.createPatient(patient);
  };
}

// TODO Attach redux functionality
