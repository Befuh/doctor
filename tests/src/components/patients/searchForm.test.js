import React from 'react';
import { shallow } from 'enzyme';
import { SearchForm } from '../../../../src/components/patients/SearchForm';

let wrapper, startUpdatePatients, preventDefault;
beforeEach(() => {
  preventDefault = jest.fn();
  startUpdatePatients = jest.fn();
  wrapper = shallow(<SearchForm startUpdatePatients={startUpdatePatients}/>);
});

describe('search form', () => {
  it('renders correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('updates the search field values', () => {
    [
      {
        name: 'firstName',
        index: 0,
        selector: 'input',
        event: 'change',
        value: 'Cristiano'
      },
      {
        name: 'lastName',
        index: 1,
        selector: 'input',
        event: 'change',
        value: 'Ronaldo'
      },
      {
        name: 'identifier',
        index: 2,
        selector: 'input',
        event: 'change',
        value: '223lslkjfÃ¶ldsjfs'
      },
      {
        name: 'sex',
        index: 0,
        selector: 'select',
        event: 'change',
        value: 'male'
      }
    ].forEach(field => {
      preventDefault = jest.fn();
      const value = field.value;
      wrapper.find(field.selector).at(field.index).simulate(field.event, { preventDefault, target: { value } });
      expect(wrapper.state()[field.name]).toEqual(field.value);
    });
  });

  it('starts the process to update the patients list', () => {
    wrapper.setState({ lastName: 'Messi' });

    wrapper.find('form').simulate('submit', { preventDefault });

    expect(preventDefault).toHaveBeenCalled();
    expect(startUpdatePatients).toHaveBeenCalledWith({
      firstName: '',
      lastName: 'Messi',
      identifier: '',
      sex: ''
    })
  });
});
