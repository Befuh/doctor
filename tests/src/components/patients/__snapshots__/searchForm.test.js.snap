// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`search form renders correctly 1`] = `
<form
  onSubmit={[Function]}
>
  <input
    onChange={[Function]}
  />
  <input
    onChange={[Function]}
  />
  <input
    onChange={[Function]}
  />
  <select
    onChange={[Function]}
  >
    <option
      value="female"
    >
      Female
    </option>
    <option
      value="male"
    >
      male
    </option>
  </select>
  <button
    type="submit"
  >
    Search
  </button>
</form>
`;
